src/main/scala/is/hail/methods/LDPrune.scala

```diff
@@ -0,0 +1,391 @@
+package is.hail.methods
+
+import java.util
+
+import is.hail.annotations.Annotation
+import is.hail.sparkextras.GeneralRDD
+import org.apache.spark.rdd.RDD
+import org.apache.spark.storage.StorageLevel
+import is.hail.sparkextras._
+import is.hail.variant._
+import is.hail.utils._
+
+object LDPrune {
+  val variantByteOverhead = 50
+  val fractionMemoryToUse = 0.25
+  val genotypesPerPack = 32
+  val nPartitionsPerCore = 3
+
+  case class GlobalPruneIntermediate(rdd: GeneralRDD[(Variant, BitPackedVector)], index: Int, persist: Boolean)
+
+  case class BitPackedVector(gs: Array[Long], nSamples: Int, mean: Double, sdRecip: Double) {
+    def unpack(): Array[Int] = {
+
+      val shifts = (0 until genotypesPerPack * 2 by 2).reverse
+
+      gs.flatMap { l =>
+        shifts.map { s => (l >> s) & 3 match {
+          case 0 => 0
+          case 1 => 1
+          case 2 => 2
+          case 3 => -1
+        }
+        }
+      }.take(nSamples)
+    }
+  }
+
+  val table = {
+    val t = Array.fill[Byte](256 * 4)(0)
+
+    (0 until 256).foreach { i =>
+      val xi = i & 3
+      val xj = (i >> 2) & 3
+      val yi = (i >> 4) & 3
+      val yj = (i >> 6) & 3
+
+      val res = findTableValue(xi, yi, xj, yj)
+
+      t(i * 4) = res._1.toByte
+      t(i * 4 + 1) = res._2.toByte
+      t(i * 4 + 2) = res._3.toByte
+      t(i * 4 + 3) = res._4.toByte
+    }
+    t
+  }
+
+  private def findTableValue(a: Int, b: Int, c: Int, d: Int): (Int, Int, Int, Int) = {
+    val r1 = findTableValue(a, b)
+    val r2 = findTableValue(c, d)
+    (r1._1 + r2._1, r1._2 + r2._2, r1._3 + r2._3, r1._4 + r2._4)
+  }
+
+  private def findTableValue(xi: Int, yi: Int): (Int, Int, Int, Int) = {
+    var xySum = 0
+    var XbarCount = 0
+    var YbarCount = 0
+    var XbarYbarCount = 0
+
+    (xi, yi) match {
+      case (3, 3) => XbarYbarCount += 1
```

Comment:
Sadly, this is likely not a jump table because they are pairs rather than ints. Therefore, yes, listing from most to least frequent will improve performance.

You could also do something like:

```
xi*10+yi match {
  case 33 => ... 
}
```

Even quicker would be a shift left by two, but then the cases don't read as nicely.